### PR TITLE
Accumulation of my work on this library

### PR DESCRIPTION
**Pull request by Ploppz**
I have been working on it for my master thesis the previous semester. At this point it is more or less finalized.
The XOR test passes the vast majority of times, but not strictly always, so I do believe it is unfitting as a test (non-deterministic), unless we give it a large number of generations.

I will not go into detail about the changes. In general I have taken inspiration from `neat-python` and how they do things. Some simplifications have been done and every possible aspect has been parameterized in `NeatParams`. You can also do hyperparameter optimization (using a Bayesian strategy through the `blackbox` crate I created).
Feel free to ask any questions about the code.

### CODE REVIEW

**Comment by Ploppz** [comment]
I don't understand from the logs why the build fails.

**Comment by TLmaK0** [comment]
Thanks @Ploppz for your help. I will try to understand all the stuff.
I think the problem with the build is the "fmt" run, could be an issue with new version. https://travis-ci.org/TLmaK0/rustneat/jobs/552707545#L637

**Comment by Ploppz** [comment]
I think I know why that test failed. In a recent commit:
~`let n_elite = if self.organisms.len() > 5 { 2 } else { 0 };`~
`let n_elite = if self.organisms.len() > 5 { 1 } else { 1 };`

If `n_elite` is `0` it is possible that the globally best organism is killed. The reason I wrote it like that (which clippy probably complained about) was to have the possibility to change these later. Will fix and see if this fixes the tests

**Comment by Aedius** [comment]
@Ploppz it's one of my commits,
i follow the comment l 122: 
` // to random selection. NOTE: For now, we always have n_elite = 2 or 0.`
you should change it accordingly

**Comment by Ploppz** [comment]
I started rebasing but I find it rather confusing and difficult due to the timespan and amount of commits. Would it be ok to just merge upstream master into this branch to fix conflicts?

**Comment by TLmaK0** [comment]
We should try to split it in smaller pull requests

File: Cargo.toml [hunk 1]
```diff
@@ -7,34 +7,48 @@ keywords = ["neuroevolution", "neat", "aumenting-topologies", "genetic", "algori
...
 serde_json = { version = "1.0", optional = true }
 
 cpython = { version = "0.1", optional = true }
 python3-sys = { version = "0.1", optional = true }
 
+slog = "2.4.1"
+slog-term = "2.4.0"
+slog-async = "2.3.0"
+
+# [dev-dependencies] # Only used for src/bin/*
```
Comment (TLmaK0) [inline Cargo.toml]: Then they should be in dev-dependencies

File: examples/openai.rs [hunk 2]
```diff
@@ -72,14 +72,15 @@ impl CartPole {
...
     #[cfg(feature = "telemetry")]
     telemetry_helper::enable_telemetry("?max_fitness=200");
 
     let mut population = Population::create_population(150);
     let mut environment = CartPole::new();
     let mut champion: Option<Organism> = None;
     while champion.is_none() {
-        population.evolve();
+        population.evolve(&mut environment, &p, true);
         population.evaluate_in(&mut environment);
```
Comment (TLmaK0) [inline examples/openai.rs]: I think "evaluate_in" is not available right now

File: examples/simple.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    while champion.is_none() && i < MAX_ITERATIONS {
+        i += 1;
+        population.evolve(&mut environment, &p, true);
+        for organism in population.get_organisms() {
+            if organism.fitness > best_fitness {
+                best_fitness = organism.fitness;
+                best_organism = Some(organism.clone());
+            }
+            if organism.fitness > 15.0 {
+                champion = Some(organism.clone());
```
Comment (TLmaK0) [inline examples/simple.rs]: Champion not used. This can return non valid organism.

**Comment by Ploppz** [comment]
Sorry for the delay. Your plan makes sense, and I will try eventually to make those pull requests. I just currently don't have a lot of time.

**Comment by TLmaK0** [comment]
Tx @Ploppz. I know this requires a big effort from you, but I think will help us to fully understand the changes.

**Pull request closed without merging**